src/storage/engine/btree_engine.rs

```diff
@@ -0,0 +1,406 @@
+    Result as EngineResult, ScanMode, Snapshot,
+};
+use storage::{CfName, Key, Value, CF_DEFAULT, CF_LOCK, CF_WRITE};
+
+type RwLockTree = RwLock<BTreeMap<Key, Value>>;
+
+#[derive(Clone)]
+pub struct BTreeEngine {
+    cf_names: Vec<CfName>,
+    cf_contents: Vec<Arc<RwLockTree>>,
+}
+
+impl BTreeEngine {
+    pub fn new(cfs: &[CfName]) -> Self {
+        let mut cf_names = vec![];
+        let mut cf_contents = vec![];
+
+        // create default cf if missing
+        if cfs.iter().find(|&&c| c == CF_DEFAULT).is_none() {
+            cf_names.push(CF_DEFAULT);
+            cf_contents.push(Arc::new(RwLock::new(BTreeMap::new())))
+        }
+
+        for cf in cfs.iter() {
+            cf_names.push(*cf);
+            cf_contents.push(Arc::new(RwLock::new(BTreeMap::new())))
+        }
+
+        Self {
+            cf_names,
+            cf_contents,
+        }
+    }
+
+    pub fn get_cf(&self, cf: CfName) -> Arc<RwLockTree> {
+        let index = self
+            .cf_names
+            .iter()
+            .position(|&c| c == cf)
+            .expect("Not exist CF!");
+        self.cf_contents[index].clone()
+    }
+}
+
+impl Default for BTreeEngine {
+    fn default() -> Self {
+        let cfs = &[CF_WRITE, CF_DEFAULT, CF_LOCK];
+        Self::new(cfs)
+    }
+}
+
+impl Engine for BTreeEngine {
+    type Iter = BTreeEngineIterator;
+    type Snap = BTreeEngineSnapshot;
+
+    fn async_write(
+        &self,
+        _ctx: &Context,
+        modifies: Vec<Modify>,
+        cb: EngineCallback<()>,
+    ) -> EngineResult<()> {
+        if modifies.is_empty() {
+            return Err(EngineError::EmptyRequest);
+        }
+        cb((CbContext::new(), write_modifies(&self, modifies)));
+
+        Ok(())
+    }
+    /// warning: It returns a fake snapshot whose content will be affected by the later modifies!
+    fn async_snapshot(&self, _ctx: &Context, cb: EngineCallback<Self::Snap>) -> EngineResult<()> {
+        cb((CbContext::new(), Ok(BTreeEngineSnapshot::new(&self))));
+        Ok(())
+    }
+}
+
+impl Debug for BTreeEngine {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "BTreeEngine []",)
+    }
+}
+
+pub struct BTreeEngineIterator {
+    tree: Arc<RwLockTree>,
+    cur_key: Option<Key>,
+    cur_value: Option<Value>,
+    valid: bool,
+    bounds: (Bound<Key>, Bound<Key>),
+}
+
+impl BTreeEngineIterator {
+    pub fn new(tree: Arc<RwLockTree>, iter_opt: IterOption) -> BTreeEngineIterator {
+        let lower_bound = match iter_opt.lower_bound() {
+            None => Unbounded,
+            Some(key) => Included(Key::from_raw(key)),
+        };
+
+        let upper_bound = match iter_opt.upper_bound() {
+            None => Unbounded,
+            Some(key) => Excluded(Key::from_raw(key)),
+        };
+        let bounds = (lower_bound.clone(), upper_bound.clone());
+        Self {
+            tree: tree.clone(),
```

Comment:
need to add comment to the function because `BTreeEngineIterator::new()` will take notable time

Reply:
sorry,I don't understand, what should I do?